llvm/lib/Support/Compression.cpp

```diff
@@ -202,6 +207,71 @@ Error zstd::decompress(ArrayRef<uint8_t> Input,
   return E;
 }
 
+static void zstdContextFree(ZSTD_DCtx *Ctx) {
+  size_t RC = ::ZSTD_freeDCtx(Ctx);
+  assert(::ZSTD_isError(RC) == 0);
+}
+
+Error zstd::decompress(ArrayRef<uint8_t> Input,
+                       SmallVectorImpl<uint8_t> &Output) {
+  size_t Res = 0;
+  std::unique_ptr<::ZSTD_DCtx, void (*)(::ZSTD_DCtx *)> Ctx(::ZSTD_createDCtx(),
+                                                            zstdContextFree);
+  Res = ::ZSTD_initDStream(&*Ctx);
+  if (::ZSTD_isError(Res))
+    return make_error<StringError>(::ZSTD_getErrorName(Res),
+                                   inconvertibleErrorCode());
+
+  // Create a buffer, this will be filled up repeatedly by
+  // ZSTD_decompressStream and appended to Output
+  SmallVector<char> Buffer;
+  Buffer.resize_for_overwrite(4096);
+
+  ::ZSTD_inBuffer InputBuffer = {
+      .src = Input.data(), .size = Input.size(), .pos = 0};
+  ::ZSTD_outBuffer OutputBuffer = {
+      .dst = Buffer.data(), .size = Buffer.size(), .pos = 0};
+
+  // First loop, read the entirety of `InputBuffer`. Stop when
+  // ZSTD_decompressedStream reports pos to be equal to size. This signals that
+  // all the data has been read and is at least in the internal state of the
+  // decompressor.
+  while (InputBuffer.pos < InputBuffer.size) {
+    Res = ::ZSTD_decompressStream(&*Ctx, &OutputBuffer, &InputBuffer);
+    if (::ZSTD_isError(Res))
+      return make_error<StringError>(::ZSTD_getErrorName(Res),
+                                     inconvertibleErrorCode());
+
+    // Check if the bufffer was filled, this might not happen in all iterations
+    // because the data read might still reside in the internal state of the
+    // decompressor
```

Comment:
Obscure comment. I've no idea.

Reply:
Basically, the decompressor might have read the data, but have decided that it wants to read more before emitting another compressed block, so `.pos` might remain `0` for the iteration.